politeiawww/invoices_test.go

```diff
@@ -0,0 +1,281 @@
+		Domain:        "Design",
+		Subdomain:     "pgui",
+		Description:   "Artwork",
+		ProposalToken: "",
+		Labor:         0,
+		Expenses:      1000,
+	}
+	lineItemMisc := cms.LineItemsInput{
+		Type:          cms.LineItemTypeMisc,
+		Domain:        "Research",
+		Subdomain:     "dcrd",
+		Description:   "reorg",
+		ProposalToken: "",
+		Labor:         0,
+		Expenses:      10000,
+	}
+
+	return []cms.LineItemsInput{
+		lineItemLabor,
+		lineItemExpense,
+		lineItemMisc,
+	}
+}
+
+func createInvoiceInput(t *testing.T, li []cms.LineItemsInput) cms.InvoiceInput {
+	t.Helper()
+
+	var (
+		month       uint = 2
+		year        uint = 2019
+		rate        uint = 4000
+		monthAvg    uint = 1651
+		name             = "test"
+		location         = "testlocation"
+		contact          = "test@gmail.com"
+		paymentaddr      = "DsUHkmH555D4tLQi5ap4gVAV86tVN29nqYi"
+	)
+
+	return cms.InvoiceInput{
+		Version:            1,
+		Month:              month,
+		Year:               year,
+		ExchangeRate:       monthAvg,
+		ContractorName:     name,
+		ContractorLocation: location,
+		ContractorContact:  contact,
+		ContractorRate:     rate,
+		PaymentAddress:     paymentaddr,
+		LineItems:          li,
+	}
+}
+
+// createInvoiceJSON creates an index file with the passed InvoiceInput
+func createInvoiceJSON(t *testing.T, ii cms.InvoiceInput) *www.File {
+	t.Helper()
+
+	file, _ := json.Marshal(ii)
+
+	return &www.File{
+		Name:    invoiceFile,
+		MIME:    mime.DetectMimeType(file),
+		Digest:  hex.EncodeToString(util.Digest(file)),
+		Payload: base64.StdEncoding.EncodeToString(file),
+	}
+}
+
+// createNewInvoice computes the merkle root of the given files, signs the
+// merkle root with the given identity then returns a NewInvoice object.
+func createNewInvoice(t *testing.T, id *identity.FullIdentity,
+	files []www.File, month uint, year uint) *cms.NewInvoice {
+
+	t.Helper()
+
+	if len(files) == 0 {
+		t.Fatalf("no files found")
+	}
+
+	// Compute merkle
+	digests := make([]*[sha256.Size]byte, 0, len(files))
+	for _, f := range files {
+		d, ok := util.ConvertDigest(f.Digest)
+		if !ok {
+			t.Fatalf("could not convert digest %v", f.Digest)
+		}
+		digests = append(digests, &d)
+	}
+	root := hex.EncodeToString(merkle.Root(digests)[:])
+
+	// Sign merkle
+	sig := id.SignMessage([]byte(root))
+
+	return &cms.NewInvoice{
+		Month:     month,
+		Year:      year,
+		Files:     files,
+		PublicKey: hex.EncodeToString(id.Public.Key[:]),
+		Signature: hex.EncodeToString(sig[:]),
+	}
+}
+
+// Invoice Validation Tests
+func TestValidateInvoice(t *testing.T) {
+	p, cleanup := newTestPoliteiawww(t, cmsWWWMode)
+	defer cleanup()
+
+	usr, id := newUser(t, p, true, false)
+
+	vli := createValidLineItems(t)
+	ii := createInvoiceInput(t, vli)
+	json := createInvoiceJSON(t, ii)
+	png := createFilePNG(t, false)
+	md := newFileRandomMD(t)
+	ni := createNewInvoice(t, id, []www.File{*json, *png}, ii.Month, ii.Year)
+
+	// Invalid signature test
+	invoiceInvalidSig := &cms.NewInvoice{
```

Comment:
Why are you saving these as pointers when you're not actually using the pointer in the test.